sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala

```diff
@@ -406,14 +415,62 @@ trait BlockingOperatorWithCodegen extends CodegenSupport {
   override def limitNotReachedChecks: Seq[String] = Nil
 }
 
+/**
+ * Leaf codegen node reading from a single RDD.
+ */
+trait InputRDDCodegen extends CodegenSupport {
+
+  def inputRDD: RDD[InternalRow]
+
+  // If the input is an RDD of InternalRow which are potentially not UnsafeRow,
+  // and there is no parent to consume it, it needs an UnsafeProjection.
+  protected val createUnsafeProjection: Boolean = (parent == null)
+
+  override def inputRDDs(): Seq[RDD[InternalRow]] = {
+    inputRDD :: Nil
+  }
+
+  override def doProduce(ctx: CodegenContext): String = {
```

Comment:
can you highlight the difference between this one and the previous `RowDataSourceScanExec.doProduce`?

Reply:
the extra `if createUnsafeProjection` check?

Reply:
The new one should be the same as the previous `RowDataSourceScanExec.doProduce` and `RDDScanExec.doProduce` if createUnsafeProjection == true, and it should be the same as the previous `InputAdapter.doProduce` and `LocalTableScanExec.doProduce` when createUnsafeProjection == false.

From the fact that `InputAdapter` was not doing an explicit unsafe projection, even though it's input could be InternalRows that are not UnsafeRows I derived an assumption that it is safe not to do so as long as there is a parent operator. This assumes that that parent operator would always result in an UnsafeProjection being eventually added, and hence the output of the WholeStageCodegen will be in UnsafeRows.

Reply:
> This assumes that that parent operator would always result in some UnsafeProjection being eventually added, and hence the output of the WholeStageCodegen unit will be UnsafeRows.

I think it's quite a hack in my patch, and that there should be some nicer interface to tell the codegened operators whether thei're dealing with UnsafeRows input, or InternalRows that may not be UnsafeRows...